File: TDD/TDD Coding Practice String Transformer/Code/Python/dummy/original/presenter.py
from trans import NONE_SELECTED_INDEX
from validator import ValidatingResult, ParamValidatingRule, Validator

class Presenter(object):
    def __init__(self, view, businesslogic):
        self.view = view
        self.businesslogic = businesslogic
        self.view.set_presenter(self)

        self.avail_selected_index = 0
        self.chain_selected_index = NONE_SELECTED_INDEX
        self.avail_transes = None
        self.chain_transes = []

    def init(self):
        self.avail_transes = self.businesslogic.get_all_transes()
        self.view.present_avail_transes(self.avail_transes)
        self.view.set_avail_selected_index(self.avail_selected_index)

    def add_trans(self):
        avail_selected_trans = self.view.get_avail_selected_trans()

        validating_result = self.validate_add_trans(avail_selected_trans)
        if validating_result.is_succeeded:
            self.chain_transes.append(avail_selected_trans)

        self.update_chain_selected_index_for_add(avail_selected_trans, validating_result.failed_reason)
        self.update_avail_selected_index_for_add(avail_selected_trans, validating_result.failed_reason)

        self.view.present_chain_transes(self.chain_transes)
        self.view.set_chain_selected_index(self.chain_selected_index)
        self.view.set_avail_selected_index(self.avail_selected_index)

    def remove_trans(self):
        chain_selected_trans = self.view.get_chain_selected_trans()

        validating_result = self.validate_remove_trans(chain_selected_trans)
        self.update_chain_selected_index_for_remove(chain_selected_trans, validating_result.failed_reason)
        if validating_result.is_succeeded:
            self.chain_transes.remove(chain_selected_trans)
        self.update_avail_selected_index_for_remove(chain_selected_trans, validating_result.failed_reason)

        self.view.present_chain_transes(self.chain_transes)
        self.view.set_avail_selected_index(self.avail_selected_index)
        self.view.set_chain_selected_index(self.chain_selected_index)

    def remove_all_transes(self):
        validating_result = self.validate_remove_all_transes()
        if validating_result.is_succeeded:
            del self.chain_transes[:]

        self.update_chain_selected_index_for_remove_all()
        self.update_avail_selected_index_for_remove_all(validating_result.failed_reason)

        self.view.present_chain_transes(self.chain_transes)
        self.view.set_chain_selected_index(self.chain_selected_index)
        self.view.set_avail_selected_index(self.avail_selected_index)

    def apply_trans_chain(self):
        source_str = self.view.get_source_str()

        validating_result = self.validate_apply_trans_chain(source_str)
        if validating_result.is_succeeded:
            self.view.present_result_str(self.businesslogic.transform(source_str, self.chain_transes))

    def validate_add_trans(self, avail_selected_trans):
        param_validating_rules = [ParamValidatingRule(avail_selected_trans,
                                                      Presenter.trans_not_specified,
                                                      self.view.notify_avail_trans_not_specified,
                                                      ValidatingResult.VRFR_AVAIL_TRANS_NOT_SPECIFIED),
                                  ParamValidatingRule(avail_selected_trans,
                                                      self.already_existed_in_chain,
                                                      self.view.notify_add_already_existed_in_chain_trans,
                                                      ValidatingResult.VRFR_ADD_ALREADY_EXISTED_IN_CHAIN_TRANS)]
        return Validator.validate(param_validating_rules)

    def update_chain_selected_index_for_add(self, avail_selected_trans, validating_result_failed_reason):
        if validating_result_failed_reason == ValidatingResult.VRFR_AVAIL_TRANS_NOT_SPECIFIED:
            return

        self.chain_selected_index = self.chain_transes.index(avail_selected_trans)

    def update_avail_selected_index_for_add(self, avail_selected_trans, validating_result_failed_reason):
        if validating_result_failed_reason == ValidatingResult.VRFR_AVAIL_TRANS_NOT_SPECIFIED:
            self.avail_selected_index = 0
            return

        selected_index = self.avail_transes.index(avail_selected_trans)
        if validating_result_failed_reason == ValidatingResult.VRFR_ADD_ALREADY_EXISTED_IN_CHAIN_TRANS:
            self.avail_selected_index = selected_index
        else:
            self.avail_selected_index = \
                0 if Presenter.is_last_index(selected_index, self.avail_transes) else selected_index + 1

    def validate_remove_trans(self, chain_selected_trans):
        param_validating_rules = [ParamValidatingRule(self.chain_transes,
                                                      Presenter.empty_list,
                                                      self.view.notify_chain_empty,
                                                      ValidatingResult.VRFR_CHAIN_EMPTY),
                                  ParamValidatingRule(chain_selected_trans,
                                                      Presenter.trans_not_specified,
                                                      self.view.notify_chain_trans_not_specified,
                                                      ValidatingResult.VRFR_CHAIN_TRANS_NOT_SPECIFIED)]
        return Validator.validate(param_validating_rules)

    def update_chain_selected_index_for_remove(self, chain_selected_trans, validating_result_failed_reason):
        if validating_result_failed_reason == ValidatingResult.VRFR_CHAIN_EMPTY:
            self.chain_selected_index = NONE_SELECTED_INDEX
            return

        if validating_result_failed_reason == ValidatingResult.VRFR_CHAIN_TRANS_NOT_SPECIFIED:
            self.chain_selected_index = 0
            return

        if len(self.chain_transes) == 1:
            self.chain_selected_index = NONE_SELECTED_INDEX
            return

        selected_index = self.chain_transes.index(chain_selected_trans)
        self.chain_selected_index = \
            0 if Presenter.is_last_index(selected_index, self.chain_transes) else selected_index

    def update_avail_selected_index_for_remove(self, chain_selected_trans, validating_result_failed_reason):
        if validating_result_failed_reason == ValidatingResult.VRFR_CHAIN_EMPTY:
            self.avail_selected_index = 0
            return

        if validating_result_failed_reason == ValidatingResult.VRFR_CHAIN_TRANS_NOT_SPECIFIED:
            return

        self.avail_selected_index = self.avail_transes.index(chain_selected_trans)

    def validate_remove_all_transes(self):
        param_validating_rules = [ParamValidatingRule(self.chain_transes,
                                                      Presenter.empty_list,
                                                      self.view.notify_chain_empty,
                                                      ValidatingResult.VRFR_CHAIN_EMPTY)]
        return Validator.validate(param_validating_rules)

    def update_chain_selected_index_for_remove_all(self):
        self.chain_selected_index = NONE_SELECTED_INDEX

    def update_avail_selected_index_for_remove_all(self, validating_result_failed_reason):
        if validating_result_failed_reason == ValidatingResult.VRFR_CHAIN_EMPTY:
            return

        self.avail_selected_index = 0

    def validate_apply_trans_chain(self, source_str):
        param_validating_rules = [ParamValidatingRule(source_str,
                                                      Presenter.empty_str,
                                                      self.view.notify_source_str_empty,
                                                      ValidatingResult.VRFR_CHAIN_EMPTY)]
        return Validator.validate(param_validating_rules)

    def already_existed_in_chain(self, trans):
        return trans in self.chain_transes

    @staticmethod
    def is_last_index(index, lst):
        return index == len(lst) - 1

    @staticmethod
    def trans_not_specified(trans):
        return trans == None

    @staticmethod
    def empty_list(lst):
        return lst == []

    @staticmethod
    def empty_str(s):
        return s == ''

File: TDD/TDD Coding Practice String Transformer/Code/Java/src/fayelab/tdd/stringtransformer/dummy/reverse/Presenter.java
package fayelab.tdd.stringtransformer.dummy.reverse;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.function.BiConsumer;
import java.util.function.Function;
import java.util.function.Supplier;

import fayelab.tdd.stringtransformer.dummy.reverse.ValidatingResult.FailedReason;

import static java.util.Arrays.asList;
import static fayelab.tdd.stringtransformer.dummy.reverse.Trans.NONE_SELECTED_INDEX;
import static fayelab.tdd.stringtransformer.dummy.reverse.Trans.NONE_SELECTED_TRANS;
import static fayelab.tdd.stringtransformer.dummy.reverse.OperData.operData;
import static fayelab.tdd.stringtransformer.dummy.reverse.ParamValidatingRule.paramValidatingRule;

public class Presenter
{
    private View view;
    private BusinessLogic businessLogic;

    private int availSelectedIndex = 0;
    private int chainSelectedIndex = NONE_SELECTED_INDEX;
    private List<String> availTranses;
    private List<String> chainTranses = new ArrayList<>();
    private String resultStr;

    public Presenter(View view, BusinessLogic businessLogic)
    {
        this.view = view;
        this.businessLogic = businessLogic;
        this.view.setPresenter(this);
    }

    public void init()
    {
        availTranses = businessLogic.getAllTranses();
        view.presentAvailTranses(availTranses);
        view.setAvailSelectedIndex(availSelectedIndex);
    }

    public void addTrans()
    {
        operTrans(operData(Optional.of(this::collectViewDataForAdd), 
                           this::buildParamValidatingRulesForAdd, 
                           this::updatePresenterDataForAdd, 
                           this::presentViewDataForBuildTransChain));
    }

    private Optional<String> collectViewDataForAdd()
    {
        return Optional.ofNullable(view.getAvailSelectedTrans());
    }

    private List<ParamValidatingRule<?>> buildParamValidatingRulesForAdd(Optional<String> viewData)
    {
        String availSelectedTrans = viewData.orElse(NONE_SELECTED_TRANS);
        return asList(
                paramValidatingRule(availSelectedTrans, Presenter::transNotSpecified, 
                        view::notifyAvailTransNotSpecified, FailedReason.AVAIL_TRANS_NOT_SPECIFIED),
                paramValidatingRule(availSelectedTrans, this::alreadyExistedInChain, 
                        view::notifyAddAlreadyExistedInChainTrans, FailedReason.ADD_ALREADY_EXISTED_IN_CHAIN_TRANS));
    }

    private void updatePresenterDataForAdd(Optional<String> viewData, ValidatingResult validatingResult)
    {
        String availSelectedTrans = viewData.orElse(NONE_SELECTED_TRANS);
        if(validatingResult.isSucceeded())
        {
            chainTranses.add(availSelectedTrans);
        }
        updateChainSelectedIndexForAdd(availSelectedTrans, validatingResult.getFailedReason());
        updateAvailSelectedIndexForAdd(availSelectedTrans, validatingResult.getFailedReason());
    }

    private void updateChainSelectedIndexForAdd(String availSelectedTrans, FailedReason failedReason)
    {
        if(failedReason == FailedReason.AVAIL_TRANS_NOT_SPECIFIED)
        {
            return;
        }

        chainSelectedIndex = chainTranses.indexOf(availSelectedTrans);
    }

    private void updateAvailSelectedIndexForAdd(String availSelectedTrans, FailedReason failedReason)
    {
        if(failedReason == FailedReason.AVAIL_TRANS_NOT_SPECIFIED)
        {
            availSelectedIndex = 0;
            return;
        }

        int selectedIndex = availTranses.indexOf(availSelectedTrans);
        if(failedReason == FailedReason.ADD_ALREADY_EXISTED_IN_CHAIN_TRANS)
        {
            availSelectedIndex = selectedIndex;
        }
        else
        {
            availSelectedIndex = isLastIndex(selectedIndex, availTranses) ? 0 : selectedIndex + 1;
        }
    }

    public void removeTrans()
    {
        operTrans(operData(Optional.of(this::collectViewDataForRemove), 
                           this::buildParamValidatingRulesForRemove, 
                           this::updatePresenterDataForRemove, 
                           this::presentViewDataForBuildTransChain));
    }

    private Optional<String> collectViewDataForRemove()
    {
        return Optional.ofNullable(view.getChainSelectedTrans());
    }

    private List<ParamValidatingRule<?>> buildParamValidatingRulesForRemove(Optional<String> viewData)
    {
        String chainSelectedTrans = viewData.orElse(NONE_SELECTED_TRANS);
        return asList(
                paramValidatingRule(chainTranses, Presenter::emptyList, 
                        view::notifyChainEmpty, FailedReason.CHAIN_EMPTY),
                paramValidatingRule(chainSelectedTrans, Presenter::transNotSpecified, 
                        view::notifyChainTransNotSpecified, FailedReason.CHAIN_TRANS_NOT_SPECIFIED));
    }

    private void updatePresenterDataForRemove(Optional<String> viewData, ValidatingResult validatingResult)
    {
        String chainSelectedTrans = viewData.orElse(NONE_SELECTED_TRANS);
        updateChainSelectedIndexForRemove(chainSelectedTrans, validatingResult.getFailedReason());
        if(validatingResult.isSucceeded())
        {
            chainTranses.remove(chainSelectedTrans);
        }
        updateAvailSelectedIndexForRemove(chainSelectedTrans, validatingResult.getFailedReason());
    }

    private void updateChainSelectedIndexForRemove(String chainSelectedTrans, FailedReason failedReason)
    {
        if(failedReason == FailedReason.CHAIN_EMPTY)
        {
            chainSelectedIndex = NONE_SELECTED_INDEX;
            return;
        }

        if(failedReason == FailedReason.CHAIN_TRANS_NOT_SPECIFIED)
        {
            chainSelectedIndex = 0;
            return;
        }

        if(chainTranses.size() == 1)
        {
            chainSelectedIndex = NONE_SELECTED_INDEX;
            return;
        }

        int selectedIndex = chainTranses.indexOf(chainSelectedTrans);
        chainSelectedIndex = isLastIndex(selectedIndex, chainTranses) ? 0 : selectedIndex;
    }

    private void updateAvailSelectedIndexForRemove(String chainSelectedTrans, FailedReason failedReason)
    {
        if(failedReason == FailedReason.CHAIN_EMPTY)
        {
            availSelectedIndex = 0;
            return;
        }

        if(failedReason == FailedReason.CHAIN_TRANS_NOT_SPECIFIED)
        {
            return;
        }

        availSelectedIndex = availTranses.indexOf(chainSelectedTrans);
    }

    public void removeAllTranses()
    {
        operTrans(operData(Optional.empty(), 
                           this::buildParamValidatingRulesForRemoveAll, 
                           this::updatePresenterDataForRemoveAll, 
                           this::presentViewDataForBuildTransChain));
    }

    private List<ParamValidatingRule<?>> buildParamValidatingRulesForRemoveAll(Optional<?> emptyViewData)
    {
        return asList(
                paramValidatingRule(chainTranses, Presenter::emptyList, 
                        view::notifyChainEmpty, FailedReason.CHAIN_EMPTY));
    }

    private void updatePresenterDataForRemoveAll(Optional<?> emptyViewData, ValidatingResult validatingResult)
    {
        if(validatingResult.isSucceeded())
        {
            chainTranses.clear();
        }
        updateChainSelectedIndexForRemoveAll();
        updateAvailSelectedIndexForRemoveAll(validatingResult.getFailedReason());
    }

    private void updateChainSelectedIndexForRemoveAll()
    {
        chainSelectedIndex = NONE_SELECTED_INDEX;
    }

    private void updateAvailSelectedIndexForRemoveAll(FailedReason failedReason)
    {
        if(failedReason == FailedReason.CHAIN_EMPTY)
        {
            return;
        }

        availSelectedIndex = 0;
    }

    public void applyTransChain()
    {
        operTrans(operData(Optional.of(this::collectViewDataForApply), 
                           this::buildParamValidatingRulesForApply, 
                           this::updatePresenterDataForApply, 
                           this::presentViewDataForApply));
    }

    private Optional<String> collectViewDataForApply()
    {
        return Optional.ofNullable(view.getSourceStr());
    }

    private List<ParamValidatingRule<?>> buildParamValidatingRulesForApply(Optional<String> viewData)
    {
        String sourceStr = viewData.get();
        return asList(
                paramValidatingRule(sourceStr, Presenter::emptyStr, 
                        view::notifySourceStrEmpty, FailedReason.SOURCE_STR_EMPTY),
                paramValidatingRule(sourceStr, Presenter::illegalSourceStr, 
                        view::notifySourceStrIllegal, FailedReason.SOURCE_STR_ILLEGAL),
                paramValidatingRule(chainTranses, Presenter::emptyList, 
                        view::notifyChainEmpty, FailedReason.CHAIN_EMPTY));
    }

    private void updatePresenterDataForApply(Optional<String> viewData, ValidatingResult validatingResult)
    {
        resultStr = validatingResult.isSucceeded() ? businessLogic.transform(viewData.get(), chainTranses) : "";
        updateAvailSelectedIndexForApply(validatingResult.getFailedReason());
    }

    private void updateAvailSelectedIndexForApply(FailedReason failedReason)
    {
        if(failedReason == FailedReason.CHAIN_EMPTY)
        {
            availSelectedIndex = 0;
        }
    }

    private void presentViewDataForApply()
    {
        view.presentResultStr(resultStr);
        view.setAvailSelectedIndex(availSelectedIndex);
    }

    private <T> void operTrans(OperData<T> operData)
    {
        Optional<T> viewData = operData.getCollectViewDataFunc().orElse(() -> Optional.empty()).get();
        ValidatingResult validatingResult = Validator.validate(operData.getBuildParamValidatingRulesFunc().apply(viewData));
        operData.getUpdatePresenterDataFunc().accept(viewData, validatingResult);
        operData.getPresentViewDataFunc().run();
    }

    private void presentViewDataForBuildTransChain()
    {
        view.presentChainTranses(chainTranses);
        view.setChainSelectedIndex(chainSelectedIndex);
        view.setAvailSelectedIndex(availSelectedIndex);
    }

    private boolean alreadyExistedInChain(String trans)
    {
        return chainTranses.contains(trans);
    }

    private static boolean isLastIndex(int index, List<?> list)
    {
        return index == list.size() - 1;
    }

    private static boolean transNotSpecified(String trans)
    {
        return trans == NONE_SELECTED_TRANS;
    }

    private static boolean emptyList(List<?> list)
    {
        return list.isEmpty();
    }

    private static boolean emptyStr(String str)
    {
        return str.isEmpty();
    }

    private static boolean illegalSourceStr(String str)
    {
        return str.matches(".*[\u4e00-\u9fa5]+.*");
    }
}

class OperData<T>
{
    private Optional<Supplier<Optional<T>>> collectViewDataFunc;
    private Function<Optional<T>, List<ParamValidatingRule<?>>> buildParamValidatingRulesFunc;
    private BiConsumer<Optional<T>, ValidatingResult> updatePresenterDataFunc;
    private Runnable presentViewDataFunc;

    static <T> OperData<T> operData(
            Optional<Supplier<Optional<T>>> collectViewDataFunc, 
            Function<Optional<T>, List<ParamValidatingRule<?>>> buildParamValidatingRulesFunc, 
            BiConsumer<Optional<T>, ValidatingResult> updatePresenterDataFunc, 
            Runnable presentViewDataFunc)
    {
        return new OperData<>(collectViewDataFunc, buildParamValidatingRulesFunc, 
                updatePresenterDataFunc, presentViewDataFunc);
    }

    private OperData(Optional<Supplier<Optional<T>>> collectViewDataFunc, 
                     Function<Optional<T>, List<ParamValidatingRule<?>>> buildParamValidatingRulesFunc, 
                     BiConsumer<Optional<T>, ValidatingResult> updatePresenterDataFunc, 
                     Runnable presentViewDataFunc)
    {
        this.collectViewDataFunc = collectViewDataFunc;
        this.buildParamValidatingRulesFunc = buildParamValidatingRulesFunc;
        this.updatePresenterDataFunc = updatePresenterDataFunc;
        this.presentViewDataFunc = presentViewDataFunc;
    }

    public Optional<Supplier<Optional<T>>> getCollectViewDataFunc()
    {
        return collectViewDataFunc;
    }

    public Function<Optional<T>, List<ParamValidatingRule<?>>> getBuildParamValidatingRulesFunc()
    {
        return buildParamValidatingRulesFunc;
    }

    public BiConsumer<Optional<T>, ValidatingResult> getUpdatePresenterDataFunc()
    {
        return updatePresenterDataFunc;
    }

    public Runnable getPresentViewDataFunc()
    {
        return presentViewDataFunc;
    }
}
